services/feegrant-indexer/feegrant-db.service.ts

```diff
@@ -0,0 +1,170 @@
+import CallApiMixin from '../../mixins/callApi/call-api.mixin';
+import { dbFeegrantMixin } from '../../mixins/dbMixinMongoose';
+import { IFeegrantData } from './feegrant-tx-handler.service';
+const QueueService = require('moleculer-bull');
+const CONTRACT_URI = Config.CONTRACT_URI;
+const MAX_RETRY_REQ = Config.ASSET_INDEXER_MAX_RETRY_REQ;
+const ACTION_TIMEOUT = Config.ASSET_INDEXER_ACTION_TIMEOUT;
+const OPTs: CallingOptions = { timeout: ACTION_TIMEOUT, retries: MAX_RETRY_REQ };
+
+export default class FeegrantDB extends Service {
+    private callApiMixin = new CallApiMixin().start();
+    private dbFeegrantMixin = dbFeegrantMixin;
+
+    public constructor(public broker: ServiceBroker) {
+        super(broker);
+        this.parseServiceSchema({
+            name: 'db-feegrant',
+            version: 1,
+            mixins: [
+                QueueService(QueueConfig.redis, QueueConfig.opts),
+                this.dbFeegrantMixin,
+                this.callApiMixin,
+            ],
+            queues: {
+                'feegrant.db': {
+                    concurrency: 1,
+                    process(job: Job) {
+                        job.progress(10);
+
+                        // @ts-ignore
+                        this.handleJob(job.data.listUpdateFeegrantDb);
+
+                        job.progress(100);
+                        return true;
+                    },
+                },
+                'feegrant-check-expire.db': {
+                    concurrency: 1,
+                    process(job: Job) {
+                        job.progress(10);
+
+                        // @ts-ignore
+                        this.handleJobCheckExpire();
+
+                        job.progress(100);
+                        return true;
+                    },
+                },
+            }
+        });
+    }
+
+    async handleJobCheckExpire() {
+        // check expired
+        await this.adapter.updateMany({
+            'timestamp': {
+                $lte: new Date()
+            },
+            'status': FEEGRANT_STATUS.AVAILABLE
+        },
+            {
+                $set: {
+                    'status': FEEGRANT_STATUS.EXPIRED
+                }
+            }
+        )
+    }
+
+    async handleJob(listUpdateFeegrantDb: FeegrantEntity[]): Promise<any[]> {
+        // for each new created record received, insert new record 
+        await Promise.all(listUpdateFeegrantDb.map(async e => {
+            if (e.action == FEEGRANT_ACTION.CREATE) {
+                let record = {
+                    ...e,
+                } as FeegrantEntity
+                this.logger.info(`Feegrant-create: ${record}`)
+                await this.adapter.insert(record)
+            }
+        }))
+        for (const e of listUpdateFeegrantDb) {
+            // for each new used record received, update spendable
+            if (e.action == FEEGRANT_ACTION.USE) {
+                const record = await this.adapter.find({
```

Comment:
Sao hay for await thế ko get all 1 lần đc à. Chỉ là query theo origin_feegrant_txhash thôi mà

Reply:
ko được anh ạ. Với mỗi element (action thay đổi feegrant), mình cần query original feegrant của element đó rồi thực hiện update. Nếu mà tìm cách optimize hơn thì cũng chỉ optimize đc vài trường hợp đặc biệt, độ phức tạp vẫn là O(n) anh ạ

Reply:
Query hết ra xong for trong mảng đấy mà xử lý ko đc à :-?